runtime/common/src/impls.rs

```diff
@@ -98,13 +100,78 @@ pub fn era_payout(
 	(staking_payout, rest)
 }
 
+/// Simple struct which contains both an XCM `location` and `asset_id` to identify an asset which
+/// exists on some chain.
+#[derive(
+	Encode, Decode, Eq, PartialEq, Clone, RuntimeDebug, scale_info::TypeInfo, MaxEncodedLen,
+)]
+pub struct LocatableAssetId {
+	/// The (relative) location in which the asset ID is meaningful.
+	pub location: MultiLocation,
+	/// The asset's ID.
+	pub asset_id: AssetId,
+}
+
+impl LocatableAssetId {
+	pub fn new(location: MultiLocation, asset_id: AssetId) -> Self {
+		LocatableAssetId { location, asset_id }
+	}
+}
+
+/// Converts the [`LocatableAssetId`] to the [`xcm_builder::LocatableAssetId`].
+pub struct LocatableAssetIdConverter;
+impl Convert<LocatableAssetId, xcm_builder::LocatableAssetId> for LocatableAssetIdConverter {
+	fn convert(a: LocatableAssetId) -> xcm_builder::LocatableAssetId {
+		xcm_builder::LocatableAssetId { asset_id: a.asset_id, location: a.location }
```

Comment:
This looks silly, why can't we just directly reuse `xcm_builder::LocatableAssetId` instead of declaring a new `LocatableAssetId` here in this module?

Reply:
I do not like that a change in an external library can change the signature of the spend dispatchable.

Reply:
I don't see why you have to prepare for this? If upstream changes, then you can think of creating a new type retaining the old fields, but this now is a bit overoptimized.

Reply:
the change to the dispatchable signature will invalidate all spend proposals which has been submitted before and still active.
and such a change might be not noticed (this we can address with a test).
but this why I made it this way.

Reply:
Here's a better idea: why not have the `spend` dispatchable directly take in `Box<VersionedMultiLocation>` and `Box<VersionedAssetId>` as parameters? This should have been the proper way of accepting versioned XCM parameters anyway.

Reply:
I like the idea with versioned types, this is probably how we should do it.
We need to combine both type into one type, since it is meant for `pallet_treasury::Config::AssetKind` type parameter. This AssetKind can be just `u32` for example.

It can be same `LocatableAssetId` but with versioned versions of MultiLocation and AssetId,
```
pub struct LocatableAssetId {
	pub location: VersionedMultiLocation,
	pub asset_id: VersionedAssetId,
}
```

or i would better do something like 
```
enum VersionedLocatableAsset {
   #[codec(index = 3)]
   V3{location: v3::MultiLocation, asset_id: v3::AssetId},
}
```
this way we can control which versions relevant right now specifically for treasury proposals.

Reply:
Hmm, so what you seem to be saying is the parameters to the `spend` extrinsic is dependent on what the concrete configured type is for `pallet_treasury::Config::AssetKind`, so if it ever changes, then the signature of the `spend` extrinsic must also necessarily change. Given that's the case, why would we then allow `AssetKind` to be configurable? It would seem to me that we should instead make it not configurable.